Xamarin.Forms.Platform.Android/CarouselView/CarouselViewRenderer.cs

```diff
@@ -0,0 +1,607 @@
+ * it will be saved to ViewModel and when the view is destroyed and recreated by ViewPager, its state will be restored.
+ * 
+ */
+
+namespace Xamarin.Forms.Platform.Android
+{
+	/// <summary>
+	/// CarouselView Renderer
+	/// </summary>
+	public class CarouselViewRenderer : ViewRenderer<CarouselView, AViews.View>
+	{
+		AViews.View nativeView;
+		ViewPager viewPager;
+        CirclePageIndicator indicators;
+		bool _disposed;
+
+		double ElementWidth;
+		//double ElementHeight;
+
+		protected override void OnElementChanged(ElementChangedEventArgs<CarouselView> e)
+		{
+			base.OnElementChanged(e);
+
+			if (Control == null)
+			{
+				// Instantiate the native control and assign it to the Control property with
+				// the SetNativeControl method (called when Height BP changes)
+			}
+
+			if (e.OldElement != null)
+			{
+				// Unsubscribe from event handlers and cleanup any resources
+
+				if (viewPager != null)
+				{
+					viewPager.PageSelected -= ViewPager_PageSelected;
+					viewPager.PageScrollStateChanged -= ViewPager_PageScrollStateChanged;
+				}
+
+				if (Element != null)
+				{
+					Element.SizeChanged -= Element_SizeChanged;
+					if (Element.ItemsSource != null && Element.ItemsSource is INotifyCollectionChanged)
+					    ((INotifyCollectionChanged)Element.ItemsSource).CollectionChanged -= ItemsSource_CollectionChanged;
+				}
+			}
+
+			if (e.NewElement != null)
+			{
+				Element.SizeChanged += Element_SizeChanged;
+
+				// Configure the control and subscribe to event handlers
+				if (Element.ItemsSource != null && Element.ItemsSource is INotifyCollectionChanged)
+					((INotifyCollectionChanged)Element.ItemsSource).CollectionChanged += ItemsSource_CollectionChanged;
+			}
+		}
+
+		async void ItemsSource_CollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
+		{
+			// NewItems contains the item that was added.
+			// If NewStartingIndex is not -1, then it contains the index where the new item was added.
+			if (e.Action == NotifyCollectionChangedAction.Add)
+			{
+                InsertPage(Element?.ItemsSource.GetItem(e.NewStartingIndex), e.NewStartingIndex);
+			}
+
+			// OldItems contains the item that was removed.
+			// If OldStartingIndex is not -1, then it contains the index where the old item was removed.
+			if (e.Action == NotifyCollectionChangedAction.Remove)
+			{
+				await RemovePage(e.OldStartingIndex);
+			}
+
+			// OldItems contains the moved item.
+            // OldStartingIndex contains the index where the item was moved from.
+			// NewStartingIndex contains the index where the item was moved to.
+			if (e.Action == NotifyCollectionChangedAction.Move)
+			{
+                var Source = ((PageAdapter)viewPager?.Adapter).Source;
+
+				if (Element != null && viewPager != null && Source != null)
+				{
+					Source.RemoveAt(e.OldStartingIndex);
+					Source.Insert(e.NewStartingIndex, e.OldItems[e.OldStartingIndex]);
```

Comment:
It seem be wrong. 
e.OldItems.Count can smaller than e.OldStartingIndex

I think , it should be e.OldItems[0]